src/shims/foreign_items.rs

```diff
@@ -363,6 +363,19 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         // shim, add it to the corresponding submodule.
         match link_name.as_str() {
             // Miri-specific extern functions
+            "miri_get_alloc_id" => {
+                let [ptr] = this.check_shim(abi, Abi::Rust, link_name, args)?;
+                let ptr = this.read_pointer(ptr)?;
+                let (alloc_id, _, _) = this.ptr_get_alloc_id(ptr)?;
+                this.write_scalar(Scalar::from_u64(alloc_id.0.get()), dest)?;
+            }
+            "miri_print_stacks" => {
+                let [id] = this.check_shim(abi, Abi::Rust, link_name, args)?;
+                let id = this.read_scalar(id)?.to_u64()?;
+                if let Some(id) = std::num::NonZeroU64::new(id) {
+                    this.print_stacks(AllocId(id))?;
+                }
+            }
```

Comment:
is this separate (and on an alloc id) so we can avoid adding more stacked borrows to things when printing the stack?

Reply:
no no, I like it. Should document that though, so people don't just `miri_print_stacks(miri_get_alloc_id(&x.y))` everywhere

Reply:
True, but where?

This whole workflow where end users declare `extern "Rust"` functions is strange to say the least. Would it make sense to have something on `crates.io` that at least provides safe wrappers around these?

Reply:
> Would it make sense to have something on `crates.io` that at least provides safe wrappers around these? Having a crate for these would definitely ease documenting them.

yea, I was wondering about that, too. Such a crate is complicated to tie to a miri version, so I'm not sure publishing it will make things easier for people if we ever change anything in that crate. But it may be better than just having documentation in markdown for these magic extern functions. idk, maybe start out with docs in this repo? It's not like such a crate would be on crates.io or godbolt, so people may have to resort to locally using extern anyway

Reply:
> True, but where?

The README has a section on these magic extern functions; they need to at least be documented there.

And then we should have a chapter in the Miri Book about how to debug Stacked Borrows things. Which book, you ask? Yes that is indeed the point.^^

> yea, I was wondering about that, too. Such a crate is complicated to tie to a miri version, so I'm not sure publishing it will make things easier for people if we ever change anything in that crate.

Yes we should have such a create! This interface does not change very often, and when it does we anyway need to do some clever updating thing like we did when changing the backtrace API. So we can add a new API to Miri, ship that, then update the crate, ship that, then maybe remove the old thing from Miri.

Which reminds me, *all* these magic functions that Miri exposes should have some kind of flag/version number parameter, to make them extensible in the future.

Reply:
This documentation is still missing

Reply:
Somehow I missed this...

Do we want to add a version mangle to these new functions?

The functions in the README aren't in alphabetical order :sweat_smile: so I guess I'll just stick the new ones at the bottom?

Reply:
I think we can add the version to everything in one go when we figure out how we want to distribute these functions to users